enclave/core/string.c

```diff
@@ -138,4 +138,22 @@ char* oe_strstr(const char* haystack, const char* needle)
     return NULL;
 }
 
+char* oe_strncpy(char* dest, const char* src, size_t n)
+{
+    char* ret = dest;
+
+    if (dest == NULL)
+        return NULL;
+
+    /* Copy at most n bytes. Terminate when src is exhausted. */
+    while (n-- && *src)
+        *dest++ = *src++;
+
+    /* If there is room left, then inject zero-terminator. */
+    if (n)
+        *dest = '\0';
+
+    return ret;
```

Comment:
@mikbras @soccerGB @CodeMonkeyLeet @paulcallen @gupta-ak 
The need to implement and review oe_strncpy is one of the reasons why corelibc is more trouble than it is worth.

This implementation of oe_strncpy  looks correct - copy at most n characters from src to dst - a _bounded string-copy_.

However this implementation is incorrect because, counterintuitively, strncpy has different semantics according to the C standard. strncpy requires that the destination be padded with null characters if the src contains less that n characters. 

From [strncpy C++ reference](http://www.cplusplus.com/reference/cstring/strncpy/)


> Copies the first num characters of source to destination. If the end of the source C string (which is signaled by a null-character) is found before num characters have been copied, **destination is padded with zeros until a total of num characters have been written to it.**

From [Ansi C Rationale](http://www.lysator.liu.se/c/rat/d11.html)

>  strncpy was initially introduced into the C library to deal with fixed-length name fields in structures such as directory entries.  Such fields are not used in the same way as strings: the trailing null is unnecessary for a maximum-length field, and setting trailing bytes for shorter names to null assures efficient field-wise comparisons.  strncpy is not by origin a ``bounded strcpy,'' and the Committee has preferred to recognize existing practice rather than alter the function to better suit it to such use. 

C libraries take great pains to implement the standard; no matter how counter-intuitive the standard is. Take for example MUSL's implementation of   which strncpy forwards to:
https://github.com/microsoft/openenclave/blob/fe0c6af59d1d3ec038c47cda52555cdf351b10f8/3rdparty/musl/musl/src/string/stpncpy.c#L10-L29

MUSL's strncpy follows the C standard by null-padding the destination bytes. Additionally, when used as a glibc replacement, it does glibc compatible optimizations.

Since crypto is a key component in writing secure code, we need to shy away from writing new code in the hot path. Instead we need to rely on battle tested code like MUSL or GLIBC (along with their security fixes).
If we start implementing libc functions, when we are already using a full fledged libc implementation like MUSL, we ought to lock down each C function we implement rigorously with tests and prove to ourselves that we are standards compliant. At this moment I don't see rigorous tests for corelibc.

Reply:
Agree with Anand here.

Reply:
+1.  If we have a POSIX name we must adhere to the standard.  strncpy as specified in the standard is dangerous since it does not guarantee null termination (as explained at the link Anand cites).   As such, if we make oe_strncpy be public, then it should probably be marked as deprecated or unsafe or something, so it will generate warnings if an app tries to use it.  strncpy_s is the safer API to use.

Reply:
This is strncpy was added for the purpose getting mbedtls built on top of corelibc, where once we starting building mbedtls on top of the musl, this direct corelibc dependency will be removed thus the concern raised will be gone because this strncpy will become no longer needed.

Reply:
Can you open an issue to keep track of removing this strncpy implementation or fixing it to be libc compliant before the next release? Otherwise, we would increase the risk of unpredictable behavior/potential security vulnerability.